### PR TITLE
Fix Result post-processing

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Required for https://github.com/Qiskit/qiskit-aer/pull/1388

Fixes an issue with how the `Result` class handles post-processing for the functions `get_statevector` and `get_unitary` if the result data is already a Statevector or Operator object.

This is needed for planned feature for the next Aer release to automatically format the result data for save instructions to the corresponding terra classes (Statevector, DensityMatrix, Operator, SuperOp, Clifford, ProbDistribution).

### Details and comments

~Because of all the unexpected cyclic dependencies already in terra I couldn't import the Statevector or Operator classes for type checking in the `postprocessing` or `result` module, so had to do a duck-type check based on the class name instead.~

### CODE REVIEW

File: qiskit/result/postprocess.py [hunk 1]
```diff
@@ -182,6 +182,12 @@ def format_statevector(vec, decimals=None):
     Returns:
         list[complex]: a list of python complex numbers.
     """
+    if type(vec).__name__ == "Statevector":
+        # Duck-type check since we can't import Statevector here
+        # without introducing cyclic dependencies.
```
Comment (jakelishman) [inline qiskit/result/postprocess.py:187]: You can do this - you can protect the import inside the method and tell `pylint` to ignore the cyclical dependency.  It's done a lot elsewhere in the codebase for things like this.  Something like

```python
def format_statevector(vec, decimals=None):
    from qiskit.quantum_info import Statevector  # pylint: disable=cyclic-import
    ...
```

Actually _having_ the cyclic dependencies isn't ideal, but the ship's sailed on that one in Qiskit.

Comment (chriseclectic) [reply qiskit/result/postprocess.py:187]: If you think this is preferable I can change to do that

Comment (jakelishman) [reply qiskit/result/postprocess.py:187]: Yeah, it's the correct way to handle this sort of thing - type checking by the name of the class is quite fragile at best.

File: qiskit/result/postprocess.py [hunk 2]
```diff
@@ -182,6 +182,13 @@ def format_statevector(vec, decimals=None):
     Returns:
         list[complex]: a list of python complex numbers.
     """
+    # pylint: disable=cyclic-import,import-outside-toplevel
```
Comment (jakelishman) [inline qiskit/result/postprocess.py:185]:
```suggestion
    # pylint: disable=cyclic-import
```

File: qiskit/result/postprocess.py [hunk 3]
```diff
@@ -210,6 +217,13 @@ def format_unitary(mat, decimals=None):
     Returns:
         list[list[complex]]: a matrix of complex numbers
     """
+    # pylint: disable=cyclic-import,import-outside-toplevel
```
Comment (jakelishman) [inline qiskit/result/postprocess.py:220]:
```suggestion
    # pylint: disable=cyclic-import
```

**Review by jakelishman — COMMENTED** [review]
(`import-outside-toplevel` is disabled in our `.pylintrc`, and Lev just went on a kick removing all "useless" suppressions)

File: qiskit/result/postprocess.py [hunk 4]
```diff
@@ -182,6 +182,13 @@ def format_statevector(vec, decimals=None):
     Returns:
         list[complex]: a list of python complex numbers.
     """
+    # pylint: disable=cyclic-import,import-outside-toplevel
+    from qiskit.quantum_info.states.statevector import Statevector
+
+    if isinstance(vec, Statevector):
+        if decimals:
+            return Statevector(np.around(vec.data, decimals=decimals))
```
Comment (jakelishman) [inline qiskit/result/postprocess.py:190]: Before you had `type(vec)`, which would propagate through subclasses.  That may still be what you want to do.

Comment (chriseclectic) [reply qiskit/result/postprocess.py:190]: This should be fine, since there are no subclasses at the moment.

**Review by jakelishman — APPROVED** [review]

**Comment by jakelishman** [comment]
Oh, actually - do you need to propagate the dimensions of the statevectors and operators?

**Comment by chriseclectic** [comment]
@jakelishman If they are N-qubit (as they are in Aer) the dimensions will be computed automatically. But if someone wrote a qudit simulator yes. I can update to handle that case.

**Review by jakelishman — APPROVED** [review]
Looks fine to me.

**Pull request merged**